src/ProbabilisticProgramming/Loops.dfy

```diff
@@ -14,73 +14,76 @@ module Loops {
+  }
+
+  ghost function WhileCutTerminatesWithFuel<A>(condition: A -> bool, body: A -> Monad.Hurd<A>, init: A, s: Rand.Bitstream): nat -> bool {
+    (fuel: nat) => !condition(WhileCut(condition, body, init, fuel)(s).0)
+  }
+
   // Definition 39 / True iff Prob(iset s | While(condition, body, a)(s) terminates) == 1
-  ghost predicate WhileTerminates<A(!new)>(condition: A -> bool, body: A -> Monad.Hurd<A>) {
-    var p := (a: A) =>
-               (s: Rand.Bitstream) => exists n :: !condition(WhileCut(condition, body, n, a)(s).0);
-    forall a :: Quantifier.AlmostSurely(p(a))
+  ghost predicate WhileTerminatesAlmostSurely<A(!new)>(condition: A -> bool, body: A -> Monad.Hurd<A>) {
+    var p := (init: A) =>
+               (s: Rand.Bitstream) => WhileCutTerminates(condition, body, init, s);
+    forall init :: Quantifier.AlmostSurely(p(init))
   }
 
-  // Theorem 38
-  function While<A>(condition: A -> bool, body: A -> Monad.Hurd<A>, init: A): (f: Monad.Hurd<A>)
-    requires WhileTerminates(condition, body)
+  // Equation (3.25)
+  ghost function While<A>(condition: A -> bool, body: A -> Monad.Hurd<A>, init: A): (f: Monad.Hurd<A>)
   {
-    assume {:axiom} false; // assume termination
-    if condition(init) then
-      Monad.Bind(body(init), (a': A) => While(condition, body, a'))
-    else
-      Monad.Return(init)
+    (s: Rand.Bitstream) =>
+      if WhileCutTerminates(condition, body, init, s)
+      then
+        var fuel := MinimalFuel(condition, body, init, s);
+        WhileCut(condition, body, init, fuel)(s)
+      else
+        // In HOL, Hurd returns `arb` here, which is not possible in Dafny.
+        (init, s)
   }
 
-  method WhileImperative<A>(condition: A -> bool, body: A -> Monad.Hurd<A>, init: A, s: Rand.Bitstream) returns (t: (A, Rand.Bitstream))
-    requires WhileTerminates(condition, body)
-    ensures While(condition, body, init)(s) == t
-    decreases *
+  ghost function MinimalFuel<A>(condition: A -> bool, body: A -> Monad.Hurd<A>, init: A, s: Rand.Bitstream): (fuel: nat)
+    requires WhileCutTerminates(condition, body, init, s)
+    ensures WhileCutTerminatesWithFuel(condition, body, init, s)(fuel)
+    ensures forall fuel': nat :: WhileCutTerminatesWithFuel(condition, body, init, s)(fuel') ==> fuel <= fuel'
   {
-    while condition(init)
-      decreases *
-    {
-      var (a, s) := body(init)(s);
-    }
-    return (init, s);
+    var fuelBound :| WhileCutTerminatesWithFuel(condition, body, init, s)(fuelBound);
+    Minimal(WhileCutTerminatesWithFuel(condition, body, init, s), fuelBound)
   }
 
-  method WhileImperativeAlternative<A>(condition: A -> bool, body: A -> Monad.Hurd<A>, init: A, s: Rand.Bitstream) returns (t: (A, Rand.Bitstream))
-    requires WhileTerminates(condition, body)
-    ensures While(condition, body, init)(s) == t
-    decreases *
+  ghost function Minimal(property: nat -> bool, bound: nat, i: nat := 0): (n: nat)
```

Comment:
Let's call this `Least`, not `Minimal` (the concepts might be the same in this context, but still). Also, let's rename `n` to `l` or `least`, and `m` to `n`.